src/lib/comms/sol-lwm2m-bs-server.c

```diff
@@ -212,67 +213,95 @@ static const struct sol_coap_resource bootstrap_request_interface = {
 
 SOL_API struct sol_lwm2m_bootstrap_server *
 sol_lwm2m_bootstrap_server_new(uint16_t port, const char **known_clients,
-    enum sol_lwm2m_security_mode sec_mode, struct sol_vector *known_psks)
+    uint16_t num_sec_modes, ...)
 {
     struct sol_lwm2m_bootstrap_server *server;
     struct sol_network_link_addr servaddr = { .family = SOL_NETWORK_FAMILY_INET6,
                                               .port = port };
     int r;
-    const char *sec_mode_str;
-    struct sol_lwm2m_security_psk *psk, *given_psk;
+    struct sol_lwm2m_security_psk **known_psks = NULL;
+    struct sol_lwm2m_security_rpk *my_rpk = NULL;
+    struct sol_blob **known_pub_keys = NULL;
+    enum sol_lwm2m_security_mode *sec_modes;
+    enum sol_socket_dtls_cipher *cipher_suites;
     uint16_t i;
+    va_list ap;
 
     SOL_LOG_INTERNAL_INIT_ONCE;
 
     SOL_NULL_CHECK(known_clients, NULL);
-    switch (sec_mode) {
-    case SOL_LWM2M_SECURITY_MODE_PRE_SHARED_KEY:
-        SOL_NULL_CHECK(known_psks, NULL);
-        sec_mode_str = "Pre-Shared Key";
-        break;
-    case SOL_LWM2M_SECURITY_MODE_RAW_PUBLIC_KEY:
-        sec_mode_str = "Raw Public Key";
-        SOL_WRN("Raw Public Key security mode is not supported yet.");
-        return NULL;
-    case SOL_LWM2M_SECURITY_MODE_CERTIFICATE:
-        sec_mode_str = "Certificate";
-        SOL_WRN("Certificate security mode is not supported yet.");
-        return NULL;
-    case SOL_LWM2M_SECURITY_MODE_NO_SEC:
-        SOL_WRN("Bootstrap Server MUST use DTLS.");
-        return NULL;
-    default:
-        SOL_WRN("Unknown DTLS Security Mode: %d", sec_mode);
-        return NULL;
+
+    va_start(ap, num_sec_modes);
+
+    cipher_suites = calloc(num_sec_modes, sizeof(enum sol_socket_dtls_cipher));
+    SOL_NULL_CHECK_GOTO(cipher_suites, err_va_list);
+
+    sec_modes = calloc(num_sec_modes, sizeof(enum sol_lwm2m_security_mode));
+    SOL_NULL_CHECK_GOTO(sec_modes, err_cipher_suites);
+
+    for (i = 0; i < num_sec_modes; i++) {
+        sec_modes[i] = va_arg(ap, enum sol_lwm2m_security_mode);
+
+        switch (sec_modes[i]) {
+        case SOL_LWM2M_SECURITY_MODE_PRE_SHARED_KEY:
```

Comment:
You should add a check here. If you find the same security mode more than once you should return an error.